content/blog/2018/appswitch/index.md

```diff
@@ -0,0 +1,81 @@
+
+To facilitate discovering services that can be accessed through service references, AppSwitch provides a _service registry_.  The registry is auto-curated and automatically kept up to date as services come and go across the cluster based on the network API that AppSwitch tracks.  Each entry in the registry consists of the IP and port where the respective service is bound.  Along with that, it includes a set of labels indicating the application to which this service belongs, the IP and port that the application passed through the socket API when creating the service, the IP and port where AppSwitch actually bound the service on the underlying host on behalf of the application etc.  In addition, applications created under AppSwitch carry a set of labels passed by the user that describe the application together with a few default system labels indicating the user that created the application and the host where the application is running etc.  These labels are all available to be expressed in the label-selector carried by a service reference.  A service in the registry can be made accessible to clients by creating a service reference.  A client would then be able to reach the service at the reference’s name (IP:port).  Now let’s look at how AppSwitch guarantees that the reference remains valid even when the target service has not yet come up.
+
+### Non-blocking requests
+
+AppSwitch leverages the semantics of the BSD socket API to ensure that service references appear valid from the perspective of clients as corresponding services come up.  When a client makes a blocking connect call to another service that has not yet come up, AppSwitch blocks the call for a certain time waiting for the target service to become live.  Since it is known that the target service is a part of the application and is expected to come up shortly, making the client block rather than returning an error such as `ECONNREFUSED` prevents the application from failing to start.  If the service doesn’t come up within time, an error is returned to the application so that framework-level mechanisms like Kubernetes crash-loop can kick in.
+
```

Comment:
Now, the discussion is going towards intra-pod container startup dependencies, and not one where the client pods and server pods are in different machines.